string[] source
string[] target
float64[] prob
